src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,447 @@
+        };
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public void beginTree(DetailAST root) {
+        variables = new ArrayDeque<>();
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final int type = ast.getType();
+        if (type == TokenTypes.DOT) {
+            visitDotToken(ast);
+        }
+        else if (type == TokenTypes.VARIABLE_DEF) {
+            visitVariableDefToken(ast);
+        }
+        else if (type == TokenTypes.IDENT) {
+            visitIdentToken(ast);
+        }
+    }
+
+    @Override
+    public void leaveToken(DetailAST ast) {
+        if (TokenUtil.isOfType(ast, SCOPES)) {
+            while (!variables.isEmpty() && variables.peek().getScope() == ast) {
+                final VariableDesc variableDesc = variables.pop();
+                if (!variableDesc.isUsed()) {
+                    final DetailAST typeAst = variableDesc.getTypeAst();
+                    log(typeAst, MSG_UNUSED_LOCAL_VARIABLE, variableDesc.name);
+                }
+            }
+        }
+    }
+
+    /**
+     * Visit ast of type {@link TokenTypes#DOT}.
+     *
+     * @param dotAst dotAst
+     */
+    private void visitDotToken(DetailAST dotAst) {
+        final boolean checkIdentToken = !TokenUtil.findFirstTokenByPredicate(dotAst,
+                        ast -> {
+                            return TokenUtil.isOfType(ast,
+                                    UNACCEPTABLE_CHILD_OF_DOT);
+                        })
+                .isPresent();
+        if (checkIdentToken) {
+            checkIdentifierAst(dotAst.findFirstToken(TokenTypes.IDENT));
+        }
+    }
+
+    /**
+     * Visit ast of type {@link TokenTypes#VARIABLE_DEF}.
+     *
+     * @param varDefAst varDefAst
+     */
+    private void visitVariableDefToken(DetailAST varDefAst) {
+        if (varDefAst.getParent().getType() != TokenTypes.OBJBLOCK) {
+            final DetailAST ident = varDefAst.findFirstToken(TokenTypes.IDENT);
+            final VariableDesc desc = new VariableDesc(ident.getText(),
+                    varDefAst.findFirstToken(TokenTypes.TYPE), findScopeOfVariable(varDefAst));
+            variables.push(desc);
+        }
+    }
+
+    /**
+     * Visit ast of type {@link TokenTypes#IDENT}.
+     *
+     * @param identAst identAst
+     */
+    private void visitIdentToken(DetailAST identAst) {
+        final DetailAST parentAst = identAst.getParent();
+        if (!TokenUtil.isOfType(parentAst, UNACCEPTABLE_PARENT_OF_IDENT)) {
+            if (parentAst.getType() == TokenTypes.METHOD_REF) {
+                if (shouldCheckMethodRefIdent(identAst)) {
+                    checkIdentifierAst(identAst);
+                }
+            }
+            else {
+                checkIdentifierAst(identAst);
```

Comment:
this call is same as one in if above, I believe it can be simplified

Reply:
done